x/autopilot/keeper/liquidstake.go

```diff
@@ -16,103 +15,141 @@ import (
+}
+
+// Submits an IBC transfer of the stToken to a non-stride zone (either back to the host zone or to a different zone)
+// The sender of the transfer is the hashed receiver of the original autopilot inbound transfer
+func (k Keeper) IBCTransferStToken(
+	ctx sdk.Context,
+	stToken sdk.Coin,
+	transferMetadata transfertypes.FungibleTokenPacketData,
+	autopilotMetadata types.StakeibcPacketMetadata,
+) error {
+	hostZone, err := k.stakeibcKeeper.GetHostZoneFromHostDenom(ctx, transferMetadata.Denom)
 	if err != nil {
 		return err
 	}
 
-	return k.IBCTransferStAsset(ctx, result.StToken, addr.String(), hostZone, packetMetadata)
-}
-
-func (k Keeper) IBCTransferStAsset(ctx sdk.Context, stAsset sdk.Coin, sender string, hostZone *stakeibctypes.HostZone, packetMetadata types.StakeibcPacketMetadata) error {
-	ibcTransferTimeoutNanos := k.stakeibcKeeper.GetParam(ctx, stakeibctypes.KeyIBCTransferTimeoutNanos)
-	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
-	channelId := packetMetadata.TransferChannel
+	// If there's no channelID specified in the packet, default to the channel on the host zone
+	channelId := autopilotMetadata.TransferChannel
 	if channelId == "" {
 		channelId = hostZone.TransferChannelId
 	}
+
+	// Generate a hashed address for the sender to the next hop,
+	// to prevent impersonation at downstream zones
+	// Note: The channel ID here is different than the one used in PFM
+	// (we use the outbound channelID, they use the inbound channelID)
+	// DOUBLE CHECK ME that it shouldn't matter
+	hashedAddress, err := types.GenerateHashedAddress(channelId, transferMetadata.Sender)
+	if err != nil {
+		return err
+	}
+
+	// First we need to bank send to the hashed address
+	originalReceiver, err := sdk.AccAddressFromBech32(transferMetadata.Receiver)
+	if err != nil {
+		return err
+	}
+	hashedSender, err := sdk.AccAddressFromBech32(hashedAddress)
+	if err != nil {
+		return err
+	}
+	if err := k.bankKeeper.SendCoins(ctx, originalReceiver, hashedSender, sdk.NewCoins(stToken)); err != nil {
+		return err
+	}
+
+	// Use the default transfer timeout of 10 minutes
+	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + transfertypes.DefaultRelativePacketTimeoutTimestamp
+
+	// Submit the transfer from the hashed address
 	transferMsg := &transfertypes.MsgTransfer{
-		SourcePort:    transfertypes.PortID,
-		SourceChannel: channelId,
-		Token:         stAsset,
-		// TODO: does this reintroduce the bug in PFM where senders can be spoofed?
-		// If so, should we instead call PFM directly to forward the packet?
-		// Or should we obfuscate the sender, making it a random address?
-		Sender:           sender,
-		Receiver:         packetMetadata.IbcReceiver,
+		SourcePort:       transfertypes.PortID,
+		SourceChannel:    channelId,
+		Token:            stToken,
+		Sender:           hashedAddress,
+		Receiver:         autopilotMetadata.IbcReceiver,
```

Comment:
I was wondering this as well, but we don't know the prefix at the time (since they could be sending it anywhere), so I don't think we can validate it. 

But if the address is invalid, they transfer will just fail and the tokens will fallback to the stride receiver so I don't think it's a concern if an invalid address is passed!

Reply:
Ah yeah that makes sense! I was thinking we might be able to get the prefix from the channel-id, but just realized the light client doesn't store prefixes. 

That logic makes sense to me, it doesn't sound like a concern then! Will mark this resolved 🙂